File: python_app/models/portfolio.py
"""Portfolio Record model - translated from PORTFLIO.cpy copybook.

Mirrors the COBOL PORT-RECORD structure with composite key
(id, account_no) and client/portfolio/financial info.
"""

from decimal import Decimal
from enum import StrEnum

from pydantic import BaseModel, Field


class ClientType(StrEnum):
    """Client type codes from 88-level values in PORTFLIO."""

    INDIVIDUAL = "I"
    CORPORATE = "C"
    TRUST = "T"


class PortfolioStatus(StrEnum):
    """Portfolio status codes from 88-level values in PORTFLIO."""

    ACTIVE = "A"
    CLOSED = "C"
    SUSPENDED = "S"


class PortfolioRecord(BaseModel):
    """Full portfolio record translated from COBOL PORT-RECORD.

    Maps to PORTFLIO.cpy copybook fields:
    - PORT-KEY (id + account_no)
    - PORT-CLIENT-INFO
    - PORT-PORTFOLIO-INFO
    - PORT-FINANCIAL-INFO
    """

    # Key fields (PORT-KEY)
    id: str = Field(max_length=8, description="PORT-ID")
    account_no: str = Field(max_length=10, description="PORT-ACCOUNT-NO")

    # Client info (PORT-CLIENT-INFO)
    client_name: str = Field(max_length=30, description="PORT-CLIENT-NAME")
    client_type: ClientType = Field(description="PORT-CLIENT-TYPE: I/C/T")

    # Portfolio info (PORT-PORTFOLIO-INFO)
    create_date: str = Field(max_length=8, description="PORT-CREATE-DATE: YYYYMMDD")
    last_maint: str = Field(default="", max_length=8, description="PORT-LAST-MAINT: YYYYMMDD")
    status: PortfolioStatus = Field(default=PortfolioStatus.ACTIVE, description="PORT-STATUS: A/C/S")

    # Financial info (PORT-FINANCIAL-INFO)
    total_value: Decimal = Field(
        default=Decimal("0"), max_digits=15, decimal_places=2, description="PORT-TOTAL-VALUE S9(13)V99"
    )
    cash_balance: Decimal = Field(
        default=Decimal("0"), max_digits=15, decimal_places=2, description="PORT-CASH-BALANCE S9(13)V99"
    )

    @property
    def composite_key(self) -> str:
        """Build the composite key matching VSAM KSDS key structure (12 bytes)."""
        return f"{self.id}{self.account_no}"

File: frontend/src/types/portfolio.ts
/**
 * Domain types for the Portfolio Master Record.
 *
 * These mirror the COBOL copybook `src/copybook/common/PORTFLIO.cpy`
 * (01 PORT-RECORD). Field names are translated from COBOL hyphenated names to
 * camelCase, but the underlying semantics, lengths, and allowed values are
 * preserved so the structure can later be mapped 1:1 to the backend
 * PORTMSTR / PORTREAD programs.
 */

/** PORT-CLIENT-TYPE (X(1)) — 88-level values from PORTFLIO.cpy. */
export type ClientType = 'I' | 'C' | 'T';

/** PORT-STATUS (X(1)) — 88-level values from PORTFLIO.cpy. */
export type PortfolioStatus = 'A' | 'C' | 'S';

export const CLIENT_TYPE_LABELS: Record<ClientType, string> = {
  I: 'Individual',
  C: 'Corporate',
  T: 'Trust',
};

export const PORTFOLIO_STATUS_LABELS: Record<PortfolioStatus, string> = {
  A: 'Active',
  C: 'Closed',
  S: 'Suspended',
};

/**
 * Monetary fields are stored as plain decimal strings (e.g. "1234567.89")
 * rather than JS numbers. PORT-TOTAL-VALUE / PORT-CASH-BALANCE are
 * `S9(13)V99 COMP-3` — up to 13 integer digits plus 2 decimals, which exceeds
 * the safe-integer range of an IEEE-754 double once scaled to cents. Keeping
 * the raw decimal string preserves COMP-3 packed-decimal precision end to end.
 */
export interface Portfolio {
  /** PORT-ID — primary key (PORT-KEY). */
  portId: string;
  /** PORT-ACCOUNT-NO. */
  accountNo: string;
  /** PORT-CLIENT-NAME. */
  clientName: string;
  /** PORT-CLIENT-TYPE. */
  clientType: ClientType;
  /** PORT-CREATE-DATE PIC 9(8) — YYYYMMDD. */
  createDate: string;
  /** PORT-LAST-MAINT PIC 9(8) — YYYYMMDD. */
  lastMaintDate: string;
  /** PORT-STATUS. */
  status: PortfolioStatus;
  /** PORT-TOTAL-VALUE — signed decimal string with 2 fraction digits. */
  totalValue: string;
  /** PORT-CASH-BALANCE — signed decimal string with 2 fraction digits. */
  cashBalance: string;
  /** PORT-LAST-USER. */
  lastUser: string;
  /** PORT-LAST-TRANS PIC 9(8). */
  lastTransId: string;
}
